js/src/parse.ts

```diff
@@ -26,24 +26,58 @@ import {
 } from "./types";
 
 const FRONTMATTER_REGEX = /^---\s*\n([\s\S]*?)\n---\s*\n([\s\S]*)$/;
+const RESERVED_METADATA_KEYWORDS: (keyof PromptMetadata)[] = [
+  "name",
+  "variant",
+  "version",
+  "description",
+  "model",
+  "tools",
+  "toolDefs",
+  "config",
+  "input",
+  "output",
+  "raw",
+  "ext",
+];
+
+const BASE_METADATA: PromptMetadata<ModelConfig> = {
+  ext: {},
+  metadata: {},
+  config: {} as ModelConfig,
+};
 
 export function parseDocument<ModelConfig = Record<string, any>>(
   source: string
 ): ParsedPrompt<ModelConfig> {
   const match = source.match(FRONTMATTER_REGEX);
-
   if (match) {
     const [, frontmatter, content] = match;
     try {
-      const metadata = parse(frontmatter) as PromptMetadata<ModelConfig>;
-      return { ...metadata, template: content.trim() };
+      const parsedMetadata = parse(frontmatter) as PromptMetadata<ModelConfig>;
+      const raw = { ...parsedMetadata };
+      const pruned: PromptMetadata<ModelConfig> = { ...BASE_METADATA };
+      const ext: PromptMetadata["ext"] = {};
+      for (const k in raw) {
+        const key = k as keyof PromptMetadata;
+        if (RESERVED_METADATA_KEYWORDS.includes(key)) {
+          pruned[key] = raw[key] as any;
+        } else if (key.includes(".")) {
+          const lastDotIndex = key.lastIndexOf(".");
+          const namespace = key.substring(0, lastDotIndex);
+          const fieldName = key.substring(lastDotIndex + 1);
+          ext[namespace] = ext[namespace] || {};
+          ext[namespace][fieldName] = raw[key];
+        }
+      }
+      return { ...pruned, raw, ext, template: content.trim() };
```

Comment:
unrelated, is it possible to parse picoschema at this step as well?
The issue I'm running into is that when loading prompts in Genkit I do `env.parse` on the file, but get raw picoschema. The picoschema is not parsed until `env.compile` (which is async, and complicates things... but not a blocker).

Reply:
oh, even worse... picoschema is not rendered until action render, not even compile:
https://github.com/google/dotprompt/blob/05124bb898ad7fde9a5ae67bbc355fc1f6af573f/js/src/environment.ts#L264

Reply:
So the issue is that the schema can depend on named schemas, which depend on schema resolvers / registered schemas, which may not be available at parse time.

I think what I should probably do it have a separate `renderMetadata` method that doesn't require input arguments but fully renders picoschema and resolves tools and whatnot...

Reply:
Yeah, compile is fine (I was able to switch to compile without issues), just need schemas resolved there. Should I take a stab (it's actively blocking me right now) or do you want to do it?